File: 백준/Silver/9012. 괄호/괄호.py
import sys

input = sys.stdin.readline
top = -1

def isEmpty():
    if top == -1:
        return 1
    else:
        return 0
if __name__ == "__main__":
    N = int(input())
    for i in range(N):
        prove = input()
        isNotVPS = False
        stack = []
        top = -1
        for j in range(len(prove)):
            if prove[j] == '(':
                stack.append('(')
                top+=1
            elif prove[j] == ')':
                    if not isEmpty() and stack[top] == '(':
                        top-=1
                        stack.pop()
                    else:
                        isNotVPS = True
        if top > -1:
            isNotVPS = True
        if isNotVPS:
            print("NO")
        else:
            print("YES")

File: 백준/Silver/9012. 괄호/괄호.cc
#include <bits/stdc++.h>
using namespace std;

int main(void) {
	ios::sync_with_stdio(false);
	cin.tie(0);
	int n;
	cin >> n;

	while (n--) {
		stack<char> S = {};
		bool trigger = true;
		string str;

		cin >> str;

		for (auto c : str) {
			if (c == '(') {
				S.push(c);
			}
			if (c == ')') {
				if (!S.empty() && S.top() == '(') S.pop();
				else {
					trigger = false;
					break;
				}
			}
		}
		
		if (S.empty() && trigger) cout << "YES" << '\n';
		else cout << "NO" << '\n';
	}
}